tlatools/org.lamport.tlatools/src/tla2sany/semantic/ErrorCode.java

```diff
@@ -0,0 +1,196 @@
+ * documented here with comments.
+ */
+public enum ErrorCode {
+
+  /**
+   * Used for internal errors that function more like assertions; should
+   * likely be replaced with thrown runtime exceptions. These errors are more
+   * artifacts of the specific method SANY uses for validation than general
+   * cases that should be standardized.
+   */
+  INTERNAL_ERROR (1, ErrorLevel.ERROR),
+
+  /**
+   * Many error checks are duplicates of earlier error checks, so the process
+   * terminates before ever reaching that point. To some degree this means
+   * the checks function more as internal errors, assertions about the state
+   * at that point in the program. This assists in programmer comprehension
+   * as the semantic checking code is very large and it is difficult to know
+   * what has or has not yet been checked at any given point. However, it is
+   * useful to tag these error checks with this error code as possible basis
+   * for future action. Note that it also says *suspected* - it is certainly
+   * possible that some parse input could evade all prior error checks and
+   * make it to the error check in question, it just hasn't yet been found.
+   * If a user manages to trigger an error with this error code, that counts
+   * as a bug in the parser that should be reified with an actual error code.
+   *
+   * Arity checks are a very common class of errors tagged with this code:
+   * the match between the arity of an operator and an application of it is
+   * checked many times in the codebase, and almost all of these checks will
+   * never be reached since the semantic checking process will have failed
+   * before those points.
+   */
+  SUSPECTED_UNREACHABLE_CHECK (2, ErrorLevel.UNDEFINED),
+
+  /**
+   * SANY contains some code handling language features that are either not
+   * part of the known language standard, being considered for removal from
+   * the language standard, or overwhelmingly likely to eventually be removed
+   * from the language standard; error-checking code for these features is
+   * tagged with this code since it is not worth expending development effort
+   * on standardizing those errors and creating tests for them. If the feature
+   * manages to - against all odds - escape purgatory and become more widely
+   * supported and used, code tagged with this error can be promoted to using
+   * standardized errors with associated tests.
+   *
+   * Subexpressions are one prominent category of language feature tagged
+   * with this error code.
+   */
+  UNSUPPORTED_LANGUAGE_FEATURE (3, ErrorLevel.UNDEFINED),
+
+  /**
+   * Standardized warnings. These should not cause a parse failure.
+   */
+  EXTENDED_MODULES_SYMBOL_UNIFICATION_AMBIGUITY (1000, ErrorLevel.WARNING),
+  INSTANCED_MODULES_SYMBOL_UNIFICATION_AMBIGUITY (1001, ErrorLevel.WARNING),
+
+  /**
+   * Standardized errors. These should cause a parse failure.
+   */
+  EXTENDED_MODULES_SYMBOL_UNIFICATION_CONFLICT (2000, ErrorLevel.ERROR),
```

Comment:
Should these error codes be disjoint from the error codes in [tlc2.output.EC.java](https://github.com/tlaplus/tlaplus/blob/master/tlatools/org.lamport.tlatools/src/tlc2/output/EC.java)?

Reply:
Ah good call, yes they should. I will change their values to an appropriate unused X000 block.

Reply:
Sorted & renamed using this bash script:
```sh
ERROR_CODE_PATH="/home/ahelwer/src/tlaplus/java-tools/tlatools/org.lamport.tlatools/src/tla2sany/semantic/ErrorCode.java"
sed -i -e "s/$1/$2/g" "$ERROR_CODE_PATH"
git add "$ERROR_CODE_PATH"
for file in *$1*.tla ; do
  sed -i -e "s/$1/$2/g" "$file" ;
  git add "$file" ;
  git mv $file ${file//$1/$2} ;
done
```